include/pybind11/pybind11.h

```diff
@@ -1676,75 +1659,100 @@ template <typename T, typename D>
 struct property_cpp_function<
     T,
     D,
-    detail::enable_if_t<detail::all_of<detail::type_uses_smart_holder_type_caster<T>,
-                                       detail::type_uses_smart_holder_type_caster<D>,
-                                       std::is_pointer<D>>::value>> {
+    detail::enable_if_t<detail::all_of<std::is_pointer<D>,
+                                       detail::both_t_and_d_use_type_caster_base<T, D>>::value>> {
 
     using drp = typename std::remove_pointer<D>::type;
 
-    template <typename PM, must_be_member_function_pointer<PM> = 0>
+    template <typename PM, detail::must_be_member_function_pointer<PM> = 0>
     static cpp_function readonly(PM pm, const handle &hdl) {
-        return cpp_function(
-            [pm](handle c_hdl) -> std::shared_ptr<drp> {
-                std::shared_ptr<T> c_sp = detail::type_caster<T>::shared_ptr_from_python(c_hdl);
-                D ptr = (*c_sp).*pm;
-                return std::shared_ptr<drp>(c_sp, ptr);
-            },
-            is_method(hdl));
+        detail::type_info *tinfo = detail::get_type_info(typeid(T), /*throw_if_missing=*/true);
+        if (tinfo->holder_enum_v == detail::holder_enum_t::smart_holder) {
+            return cpp_function(
+                [pm](handle c_hdl) -> std::shared_ptr<drp> {
+                    std::shared_ptr<T> c_sp = detail::type_caster<
+                        std::shared_ptr<T>>::shared_ptr_with_responsible_parent(c_hdl);
+                    D ptr = (*c_sp).*pm;
+                    return std::shared_ptr<drp>(c_sp, ptr);
+                },
+                is_method(hdl));
+        }
+        return cpp_function([pm](const T &c) -> const D & { return c.*pm; }, is_method(hdl));
     }
 
-    template <typename PM, must_be_member_function_pointer<PM> = 0>
+    template <typename PM, detail::must_be_member_function_pointer<PM> = 0>
     static cpp_function read(PM pm, const handle &hdl) {
         return readonly(pm, hdl);
     }
 
-    template <typename PM, must_be_member_function_pointer<PM> = 0>
+    template <typename PM, detail::must_be_member_function_pointer<PM> = 0>
     static cpp_function write(PM pm, const handle &hdl) {
-        return cpp_function([pm](T &c, D value) { c.*pm = std::forward<D>(value); },
-                            is_method(hdl));
+        detail::type_info *tinfo = detail::get_type_info(typeid(T), /*throw_if_missing=*/true);
+        if (tinfo->holder_enum_v == detail::holder_enum_t::smart_holder) {
+            return cpp_function([pm](T &c, D value) { c.*pm = std::forward<D>(value); },
```

Comment:
std::move

Reply:
Done: 99b657221467a985cc26c7b5619c65126eff06e9

Thanks for catching this!